File: bindings/python/insight/indexing.py
"""Indexing and sorting operations.

Functions:
    take, nonzero, flatnonzero, argsort, sort
    masked_select, searchsorted, unique, topk
    gather, scatter, scatter_add, scatter_reduce
    interp, indices, ix_

Types:
    UniqueResult — Result of unique() with values, indices, inverse, counts
"""

from __future__ import annotations

from typing import List, Optional, Union

from insight._insight import Array, Shape  # noqa: F401
from insight._insight import (
    take as _native_take,
    nonzero as _native_nonzero,
    flatnonzero as _native_flatnonzero,
    argsort as _native_argsort,
    sort as _native_sort,
    masked_select as _native_masked_select,
    searchsorted as _native_searchsorted,
    unique as _native_unique,
    UniqueResult,
    topk as _native_topk,
    gather as _native_gather,
    scatter as _native_scatter,
    scatter_add as _native_scatter_add,
    scatter_reduce as _native_scatter_reduce,
    interp as _native_interp,
    indices as _native_indices,
    ix_ as _native_ix_,
)


def _to_shape(s: Union[list, tuple, Shape]) -> Shape:
    """Convert a list/tuple to Shape if needed."""
    if isinstance(s, Shape):
        return s
    return Shape(s)


def take(
    x: "Array",
    indices: "Array",
    axis: Optional[int] = None,
) -> "Array":
    """Take elements from an array at the given indices.

    Args:
        x: Input array.
        indices: Array of indices to take.
        axis: Axis along which to take elements. If None, the input
            array is flattened first.

    Returns:
        Array containing the taken elements.
    """
    return _native_take(x, indices, axis)


def nonzero(x: "Array") -> "Array":
    """Return the indices of non-zero elements.

    Args:
        x: Input array.

    Returns:
        Array of shape ``(ndim, n)`` where ``n`` is the number of
        non-zero elements. Each row contains the indices along the
        corresponding axis.
    """
    return _native_nonzero(x)


def flatnonzero(x: "Array") -> "Array":
    """Return the flattened indices of non-zero elements.

    Args:
        x: Input array.

    Returns:
        1-D array of indices (into the flattened array) where the
        elements are non-zero.
    """
    return _native_flatnonzero(x)


def argsort(
    x: "Array",
    axis: int = -1,
    descending: bool = False,
) -> "Array":
    """Return the indices that would sort an array.

    Args:
        x: Input array.
        axis: Axis along which to sort (default: -1, the last axis).
        descending: If True, sort in descending order.

    Returns:
        Array of indices (dtype: int64) that would sort the input.
    """
    return _native_argsort(x, axis, descending)


def sort(
    x: "Array",
    axis: int = -1,
    descending: bool = False,
) -> "Array":
    """Sort an array along the given axis.

    Args:
        x: Input array.
        axis: Axis along which to sort (default: -1, the last axis).
        descending: If True, sort in descending order.

    Returns:
        Sorted array.
    """
    return _native_sort(x, axis, descending)


def masked_select(x: "Array", mask: "Array") -> "Array":
    """Select elements where the mask is True.

    Args:
        x: Input array.
        mask: Boolean mask array (broadcastable to the shape of ``x``).

    Returns:
        1-D array of selected elements.
    """
    return _native_masked_select(x, mask)


def searchsorted(
    x: "Array",
    v: "Array",
    side: str = "left",
) -> "Array":
    """Find indices where elements should be inserted to maintain order.

    Args:
        x: Sorted 1-D input array.
        v: Values to search for.
        side: If ``'left'`` (default), returns the first suitable
            insertion index. If ``'right'``, returns the last.

    Returns:
        Array of insertion indices (same shape as ``v``).
    """
    return _native_searchsorted(x, v, side)


def unique(
    x: "Array",
    return_indices: bool = False,
    return_inverse: bool = False,
    return_counts: bool = False,
) -> "UniqueResult":
    """Return the unique elements of an array.

    Args:
        x: Input array.
        return_indices: If True, also return the indices of the first
            occurrence of each unique value.
        return_inverse: If True, also return the indices to reconstruct
            the original array from the unique values.
        return_counts: If True, also return the count of each unique
            value.

    Returns:
        A ``UniqueResult`` object with attributes:
            - ``unique``: Array of unique elements.
            - ``indices``: (optional) Indices of first occurrences.
            - ``inverse``: (optional) Indices to reconstruct input.
            - ``counts``: (optional) Count of each unique element.
    """
    return _native_unique(x, return_indices, return_inverse, return_counts)


def topk(
    x: "Array",
    k: int,
    axis: int = -1,
    largest: bool = True,
    sorted: bool = True,
) -> tuple:
    """Return the k largest (or smallest) elements and their indices.

    Args:
        x: Input array.
        k: Number of top elements to return.
        axis: Axis along which to find the top elements (default: -1).
        largest: If True (default), return the largest elements. If
            False, return the smallest.
        sorted: If True (default), return the results in sorted order.

    Returns:
        Tuple of ``(values, indices)``.
    """
    return _native_topk(x, k, axis, largest, sorted)


def gather(
    x: "Array",
    dim: int,
    index: "Array",
) -> "Array":
    """Gather values along an axis using indices.

    Collects values from ``x`` at the positions specified by ``index``
    along the given dimension.

    Args:
        x: Input array.
        dim: Axis along which to gather.
        index: Array of indices (must have the same number of
            dimensions as ``x``).

    Returns:
        Array with the same shape as ``index``, containing the
        gathered values.
    """
    return _native_gather(x, dim, index)


def scatter(
    x: "Array",
    dim: int,
    index: "Array",
    src: "Array",
) -> "Array":
    """Scatter values into an array at the given indices.

    Writes values from ``src`` into the output at positions specified
    by ``index`` along dimension ``dim``.

    Args:
        x: Input array (defines the output shape).
        dim: Axis along which to scatter.
        index: Array of indices.
        src: Source values to scatter.

    Returns:
        Array with the same shape as ``x``, with values from ``src``
        placed at the specified indices.
    """
    return _native_scatter(x, dim, index, src)


def scatter_add(
    x: "Array",
    dim: int,
    index: "Array",
    src: "Array",
) -> "Array":
    """Scatter-add values into an array at the given indices.

    Accumulates (adds) values from ``src`` into the output at positions
    specified by ``index`` along dimension ``dim``.

    Args:
        x: Input array (defines the output shape).
        dim: Axis along which to scatter.
        index: Array of indices.
        src: Source values to add.

    Returns:
        Array with the same shape as ``x``, with values from ``src``
        accumulated at the specified indices.
    """
    return _native_scatter_add(x, dim, index, src)


def scatter_reduce(
    x: "Array",
    dim: int,
    index: "Array",
    src: "Array",
    reduce: str = "replace",
) -> "Array":
    """Scatter values into an array with a reduction operation.

    Args:
        x: Input array (defines the output shape).
        dim: Axis along which to scatter.
        index: Array of indices.
        src: Source values.
        reduce: Reduction mode. One of ``'replace'``, ``'add'``,
            ``'mul'``, ``'max'``, ``'min'``.

    Returns:
        Array with the same shape as ``x``, with values from ``src``
        reduced at the specified indices.
    """
    return _native_scatter_reduce(x, dim, index, src, reduce)


def interp(
    x: "Array",
    xp: "Array",
    fp: "Array",
    left: Optional[float] = None,
    right: Optional[float] = None,
) -> "Array":
    """One-dimensional linear interpolation.

    Returns the interpolated values at ``x`` given data points
    ``(xp, fp)``.

    Args:
        x: The x-coordinates at which to evaluate the interpolated
            values.
        xp: The x-coordinates of the data points (must be increasing).
        fp: The y-coordinates of the data points (same length as
            ``xp``).
        left: Value to return for ``x < xp[0]``. If None, defaults to
            ``fp[0]``.
        right: Value to return for ``x > xp[-1]``. If None, defaults
            to ``fp[-1]``.

    Returns:
        Interpolated values, same shape as ``x``.
    """
    return _native_interp(x, xp, fp, left, right)


def indices(shape: "Shape", sparse: bool = False) -> "Array":
    """Return an array representing a grid of indices.

    Args:
        shape: Shape of the grid.
        sparse: If True, return a list of 1-D arrays (one per
            dimension) instead of a single stacked array.

    Returns:
        Array of shape ``(ndim, *shape)`` where the first dimension
        indexes the coordinate axis.
    """
    return _native_indices(_to_shape(shape), sparse)


def ix_(arrays: List["Array"]) -> List["Array"]:
    """Construct an open mesh from multiple sequences.

    Given 1-D arrays, returns broadcastable arrays that can be used
    for advanced indexing.

    Args:
        arrays: List of 1-D arrays.

    Returns:
        List of arrays, each with ``ndim == len(arrays)``, suitable for
        use in advanced indexing expressions.
    """
    return _native_ix_(arrays)


__all__ = [
    "take",
    "nonzero",
    "flatnonzero",
    "argsort",
    "sort",
    "masked_select",
    "searchsorted",
    "unique",
    "UniqueResult",
    "topk",
    "gather",
    "scatter",
    "scatter_add",
    "scatter_reduce",
    "interp",
    "indices",
    "ix_",
]

File: bindings/lua/insight/indexing.lua
--- Indexing and searching operations.
--
-- Provides functions for taking, sorting, gathering, scattering,
-- and searching elements in arrays.
--
-- @module insight.indexing

local native = require("_insight")
local M = {}

--- Take elements from an array along an axis by index.
-- @tparam Array x Input array.
-- @tparam Array indices Indices of elements to take.
-- @int[opt] axis Axis along which to take. If nil, operates on flattened array.
-- @treturn Array Gathered elements.
function M.take(x, indices, axis)
  return native.take(x, indices, axis)
end

--- Return indices of non-zero elements.
-- @tparam Array x Input array.
-- @treturn Array Indices of non-zero elements.
function M.nonzero(x)
  return native.nonzero(x)
end

--- Return indices of non-zero elements (flattened).
-- @tparam Array x Input array.
-- @treturn Array Flat indices of non-zero elements.
function M.flatnonzero(x)
  return native.flatnonzero(x)
end

--- Return indices that would sort the array.
-- @tparam Array x Input array.
-- @int[opt=-1] axis Axis along which to sort. -1 is last axis.
-- @treturn Array Index array for sorting.
function M.argsort(x, axis)
  return native.argsort(x, axis or -1)
end

--- Return a sorted copy of the array.
-- @tparam Array x Input array.
-- @int[opt=-1] axis Axis along which to sort. -1 is last axis.
-- @treturn Array Sorted array.
function M.sort(x, axis)
  return native.sort(x, axis or -1)
end

--- Return unique elements of the array.
-- @tparam Array x Input array.
-- @treturn UniqueResult Table with fields: values, indices, inverse, counts.
function M.unique(x)
  return native.unique(x)
end

--- Return the k largest or smallest elements.
-- @tparam Array x Input array.
-- @int k Number of elements to return.
-- @int[opt=-1] axis Axis along which to search. -1 is last axis.
-- @bool[opt=true] largest If true, return largest; otherwise smallest.
-- @treturn Array Top-k values.
function M.topk(x, k, axis, largest)
  return native.topk(x, k, axis or -1, largest == nil and true or largest)
end

--- Gather values along an axis by index.
-- @tparam Array x Input array.
-- @tparam Array indices Indices to gather.
-- @int[opt=0] axis Axis along which to gather.
-- @treturn Array Gathered values.
function M.gather(x, indices, axis)
  return native.gather(x, indices, axis or 0)
end

--- Scatter values into an array by index.
-- @tparam Array x Input array.
-- @tparam Array indices Indices at which to scatter.
-- @tparam Array src Source values to scatter.
-- @int[opt=0] axis Axis along which to scatter.
-- @treturn Array Updated array.
function M.scatter(x, indices, src, axis)
  return native.scatter(x, indices, src, axis or 0)
end

--- Scatter-add values into an array by index.
-- @tparam Array x Input array.
-- @tparam Array indices Indices at which to add.
-- @tparam Array src Source values to add.
-- @int[opt=0] axis Axis along which to scatter-add.
-- @treturn Array Updated array.
function M.scatter_add(x, indices, src, axis)
  return native.scatter_add(x, indices, src, axis or 0)
end

--- Scatter with reduction into an array by index.
-- @tparam Array x Input array.
-- @tparam Array indices Indices at which to reduce.
-- @tparam Array src Source values.
-- @string reduce Reduction type ("add", "mul", "mean", "amax", "amin").
-- @int[opt=0] axis Axis along which to scatter-reduce.
-- @treturn Array Updated array.
function M.scatter_reduce(x, indices, src, reduce, axis)
  return native.scatter_reduce(x, indices, src, reduce, axis or 0)
end

--- 1-D linear interpolation.
-- @tparam Array x Points at which to evaluate.
-- @tparam Array xp Known x-coordinates (must be sorted).
-- @tparam Array fp Known y-coordinates.
-- @treturn Array Interpolated values.
function M.interp(x, xp, fp)
  return native.interp(x, xp, fp)
end

--- Return an array of indices (meshgrid-like).
-- @tparam table dimensions Shape of the output grid.
-- @treturn Array Index arrays.
function M.indices(dimensions)
  return native.indices(dimensions)
end

--- Construct open mesh from sequences (for advanced indexing).
-- @tparam table args Sequence of 1-D arrays.
-- @treturn table List of open mesh arrays.
function M.ix_(...)
  return native.ix_(...)
end

return M
